presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java

```diff
@@ -68,6 +72,10 @@
     private final AtomicLong totalRowsAdded = new AtomicLong();
     private final AtomicLong totalBufferedPages = new AtomicLong();
 
+    private final ConcurrentMap<Lifespan, AtomicLong> outstandingPageCountPerLifespan = new ConcurrentHashMap<>();
+    private final Set<Lifespan> noMorePagesForLifespan = ConcurrentHashMap.newKeySet();
+    private Consumer<Lifespan> lifespanCompletionCallback;
```

Comment:
`volatile`. I don't see an ordering issue in current database. `SqlQueryExecution` is published under the `synchronized` block. But just to be safe.